File: src/com/fate/engine/math/Vector3f.java
package com.fate.engine.math;

public class Vector3f {
	private float x, y, z;
	public Vector3f(float x, float y, float z) {
		this.x = x;
		this.y = y;
		this.z = z;
	}
	
	/**
	 * Clones a Vector3f.
	 */
	public Vector3f(Vector3f v) {
		this(v.getX(), v.getY(), v.getZ());
	}
	
	/**
	 * Creates a new Vector3f from a Vector2f and a supplied z value.
	 * @param v Vector2f containing new x and y values.
	 */
	public Vector3f(Vector2f v, float z) {
		this(v.getX(), v.getY(), z);
	}
	
	/**
	 * Creates a new Vector3f from a Vector2f by using the x and y values of the Vector2f.
	 * @param v Vector2f containing new x and y values.
	 */
	public Vector3f(Vector2f v) {
		this(v, 0);
	}
	
	public Vector3f() {
		this.x = 0;
		this.y = 0;
		this.z = 0;
	}
	
	public String toString() {
		return "Vector3f(" + x + ", " + y + ", " + z + ")";
	}
	
	public Matrix4f asTranslationMatrix() {
		return new Matrix4f(
				1, 0, 0, x,
				0, 1, 0, y,
				0, 0, 1, z,
				0, 0, 0, 1
		);
	}
	
	public Matrix4f asScaleMatrix() {
		return new Matrix4f(
				x, 0, 0, 0,
				0, y, 0, 0,
				0, 0, z, 0,
				0, 0, 0, 1
		);
	}
	
	public Vector3f add(Vector3f v) {
		return new Vector3f(x + v.getX(), y + v.getY(), z + v.getZ());
	}
	
	public Vector3f sub(Vector3f v) {
		return new Vector3f(x - v.getX(), y - v.getY(), z - v.getZ());
	}
	
	public Vector3f scale(float k) {
		return new Vector3f(x * k, y * k, z * k);
	}
	
	public Vector3f invScale(float k) {
		return new Vector3f(x / k, y / k, z / k);
	}
	
	public Vector3f cross(Vector3f v) {
		return new Vector3f(y * v.getZ() - z * v.getY(), z * v.getX() - x * v.getZ(), x * v.getY() - y * v.getX());
	}

	public float getX() {
		return x;
	}
	
	public float getY() {
		return y;
	}
	
	public float getZ() {
		return z;
	}
	
	public void setX(float x) {
		this.x = x;
	}
	
	public void setY(float y) {
		this.y = y;
	}
	
	public void setZ(float z) {
		this.z = z;
	}
	
	public void set(float x, float y, float z) {
		this.x = x;
		this.y = y;
		this.z = z;
	}
	
	public void set(Vector3f v) {
		this.x = v.getX();
		this.y = v.getY();
		this.z = v.getZ();
	}

	public Vector2f getXY() {
		return new Vector2f(x, y);
	}

	public float getLength() {
		return (float) Math.sqrt(x*x + y*y + z*z);
	}

	public Vector3f negate() {
		return new Vector3f(-x, -y, -z);
	}
}

File: src/main/kotlin/com/morph/engine/math/Vector3f.kt
package com.morph.engine.math

import java.util.function.Function

class Vector3f {
    var x: Float = 0.toFloat()
    var y: Float = 0.toFloat()
    var z: Float = 0.toFloat()

    val xy: Vector2f
        get() = Vector2f(x, y)

    val length: Float
        get() = Math.sqrt((x * x + y * y + z * z).toDouble()).toFloat()

    constructor(x: Float, y: Float, z: Float) {
        this.x = x
        this.y = y
        this.z = z
    }

    /**
     * Clones a Vector3f.
     */
    constructor(v: Vector3f) : this(v.x, v.y, v.z) {}

    /**
     * Creates a new Vector3f from a Vector2f and a supplied z value.
     * @param v Vector2f containing new x and y values.
     */
    @JvmOverloads constructor(v: Vector2f, z: Float = 0f) : this(v.x, v.y, z) {}

    constructor() {
        this.x = 0f
        this.y = 0f
        this.z = 0f
    }

    override fun toString(): String {
        return "Vector3f($x, $y, $z)"
    }

    fun asTranslationMatrix(): Matrix4f {
        return Matrix4f(
                1f, 0f, 0f, x,
                0f, 1f, 0f, y,
                0f, 0f, 1f, z,
                0f, 0f, 0f, 1f
        )
    }

    fun asScaleMatrix(): Matrix4f {
        return Matrix4f(
                x, 0f, 0f, 0f,
                0f, y, 0f, 0f,
                0f, 0f, z, 0f,
                0f, 0f, 0f, 1f
        )
    }

    fun add(v: Vector3f): Vector3f {
        return Vector3f(x + v.x, y + v.y, z + v.z)
    }

    fun sub(v: Vector3f): Vector3f {
        return Vector3f(x - v.x, y - v.y, z - v.z)
    }

    fun scale(k: Float): Vector3f {
        return Vector3f(x * k, y * k, z * k)
    }

    fun invScale(k: Float): Vector3f {
        return Vector3f(x / k, y / k, z / k)
    }

    fun cross(v: Vector3f): Vector3f {
        return Vector3f(y * v.z - z * v.y, z * v.x - x * v.z, x * v.y - y * v.x)
    }

    fun dot(v: Vector3f): Float {
        return x * v.x + y * v.y + z * v.z
    }

    operator fun set(x: Float, y: Float, z: Float) {
        this.x = x
        this.y = y
        this.z = z
    }

    fun set(v: Vector3f) {
        this.x = v.x
        this.y = v.y
        this.z = v.z
    }

    fun negate(): Vector3f {
        return Vector3f(-x, -y, -z)
    }

    fun map(`fun`: Function<Float, Float>): Vector3f {
        return Vector3f(`fun`.apply(x), `fun`.apply(y), `fun`.apply(z))
    }

    companion object {
        fun reflect(n: Vector3f, v: Vector3f): Vector3f {
            return v.add(n.scale(2 * v.dot(n)))
        }
    }
}
